src/experiments/74.particles-emitter.tsx

```diff
@@ -0,0 +1,192 @@
+      window.removeEventListener('touchmove', updateTouchPosition)
+    }
+  }, [])
+
+  const color = useMemo(
+    () => hexToRgb(particleColor) || '187, 134, 255',
+    [particleColor]
+  )
+
+  /* Particles */
+  useEffect(() => {
+    const container = containerRef.current
+    if (!container) return
+
+    let frameCount = 0
+
+    const renderParticles = () => {
+      frameCount++
+
+      if (frameCount % updateFrequency === 0) {
+        let shadow = ''
+
+        particles.current = particles.current.filter((p, i) => {
+          // gravity
+          if (gravityStrength > 0) p.speedY += gravityStrength * 0.98
+
+          const dx = mousePosition.current.x - p.x
+          const dy = mousePosition.current.y - p.y
+          const distance = Math.sqrt(dx * dx + dy * dy)
+
+          if (gravityStrength < 0) {
+            // push away
+            p.speedX -= (dx / distance) * updateFrequency
+            p.speedY -= (dy / distance) * updateFrequency
+          } else {
+            // pull towards mouse
+            p.speedX += (dx / distance) * updateFrequency
+            p.speedY += (dy / distance) * updateFrequency
+          }
+
+          // particle position
+          p.x += p.speedX
+          p.y += p.speedY
+
+          p.size = Math.max(
+            p.size - particleSize / particleLifespan / updateFrequency,
+            0
+          )
+          p.opacity = Math.max(
+            p.opacity - 1 / particleLifespan / updateFrequency,
+            0
+          )
+
+          if (i > 0) shadow += ', '
+          shadow += `${p.x - p.size}px ${p.y}px 0px ${
+            p.size
+          }px rgba(${color}, ${p.opacity})`
+
+          return p.size > 0 && p.opacity > 0
+        })
+
+        if (
+          particles.current.length < maxParticles &&
+          Math.random() < particleCreationRate / updateFrequency
+        ) {
+          const angle = Math.random() * Math.PI * 2
+
+          let radius = Math.PI * Math.pow(emissionRadius, 2) * gravityStrength
+
+          if (gravityStrength < 0) radius *= -1
+
+          particles.current.push({
+            x: mousePosition.current.x + Math.cos(angle) * radius,
+            y: mousePosition.current.y + Math.sin(angle) * radius,
+            size: particleSize,
+            speedX: (Math.random() - 0.5) * 0.5 * radius,
+            speedY: (Math.random() - 0.5) * 0.5 * radius,
+            opacity: 1,
+            color: `rgba(${color}, 1)`
+          })
+        }
+
+        container.style.boxShadow = shadow
+      }
+
+      animationRef.current = requestAnimationFrame(renderParticles)
+    }
+
+    renderParticles()
+
+    return () => {
+      if (animationRef.current) {
+        cancelAnimationFrame(animationRef.current)
+      }
+    }
+  }, [
+    particleCreationRate,
+    particleSize,
+    color,
+    gravityStrength,
+    particleLifespan,
+    emissionRadius,
+    maxParticles,
+    updateFrequency
```

Comment:
do we need all of this to re-define the useEffect?

I understand in a normal situation we would do it like this but since we have a slide is killing performance when I change them

lets use a useStateToRef to make all of this variables into a single ref:
```ts
import type { MutableRefObject } from 'react';
import { useEffect, useRef } from 'react';

export function useStateToRef<T = unknown>(initial: T): MutableRefObject<T> {
  const ref = useRef<T>(initial);

  useEffect(() => {
    ref.current = initial;
  }, [initial]);

  return ref;
}

```

then we can just do

```ts
const configRef = useStateToRef({
    particleCreationRate,
    particleSize,
    color,
    gravityStrength,
    particleLifespan,
    emissionRadius,
    maxParticles,
    updateFrequency
})

useEffect(
  ()=> {
  
    const {particleCreationRate, ...etc} = configRef.current
  
  },
  [configRef]
)

```